KAFKA-5879; Controller should read the latest IsrChangeNotification znodes when handling IsrChangeNotification event

lindong28 opened the pull request:
…

lindong28 committed:
KAFKA-5879; Controller should read the latest IsrChangeNotification znodes when handling IsrChangeNotification event

ijuma commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1205,6 +1205,8 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, time: Time, met
     override def process(): Unit = {
       if (!isActive) return
+
+      info(s"Process broker change event with broker list ${brokersOnEventReceived.mkString(",")}")
```
If we want to log each event start, we can do that generically in `ControllerEventManager`.

lindong28 replied:
@ijuma I think the current approach may be better than adding a toString() method to the `ControllerEvent` trait for these reasons:

- We may want to log something that is only available when we process a event. For example, when controller processes a `IsrChangeNotification` event, we may want to log the current list of sequence number under IsrChangeNotificationPath, not the list of sequence number when this event is enqueued.

- We may not be interested in logging every event in the controller log. For example, we probably don't need to log `LeaderAndIsrResponseReceived`. It is possible to let `LeaderAndIsrResponseReceived.toString()` return an empty string or None, and let `ControllerEventThread` skip the log if the string is empty or None. But that seems a bit unnecessary.

After thinking about the two approaches, I kind of prefer the current approach because we need its flexibility to decide which event to print and log arbitrary information during the event processing. What do you think?

ijuma replied:
Generally, flexible approaches lead to inconsistency and that log you really need is missing when you need it because someone forgot to add it. For the generic approach, you can log when it's enqueued and/or when it's about to be executed, depending on what you want. I'd personally prefer to try that approach first and see if we can make it work well. cc @junrao and @onurkaraman for additional opinions.

lindong28 replied:
Sorry, I forgot about this comment and haven't had the chance to look into this. In the interest of finishing this for 1.0 release, I have updated the patch to remove unnecessary log statements.

ijuma commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1366,16 +1370,17 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, time: Time, met
     def state = ControllerState.IsrChange
 
     override def process(): Unit = {
+      val currentSequenceNumbers = zkUtils.getChildrenParentMayNotExist(ZkUtils.IsrChangeNotificationPath)
```
We'd want to add a comment explaining why this is important. Similar to how we did for the other event where we applied a similar pattern.

lindong28 replied:
Sure. I added the following comment:

```
// Read the current isr change notification znodes from ZK again instead of using sequenceNumbers
// to increase the odds of processing recent isr changes in a single ControllerEvent
// and to reduce the odds of trying to access znodes that have already been deleted (KAFKA-5879).
```

ijuma commented:
@junrao, @onurkaraman, is this something we need for 1.0.0?

onurkaraman commented:
This PR seems like a temporary fix for a controller that still uses ZkClient.

The issue will go away once https://github.com/apache/kafka/pull/3765 gets checked in. Once we start using the async ZookeeperClient, we'll always lookup zk at event-processing-time instead of event-creation-time.

ijuma commented:
Thanks. We should decide whether we need a fix for 1.0.0, which won't have the async ZookeeperClient then.

lindong28 commented:
@ijuma @onurkaraman Based on my experience, this issue effectively prevents Kafka cluster from reassigning 200+ partitions at a time because the reassignment is too low due to huge amounts of ZK reads. This impacts user experience and I think we should fix this issue in 1.0. This patch is needed if we won't include #3765 in 1.0 release.

junrao commented:
@lindong28 : I thought that we batch the isr change propagation already in ReplicaManager.maybePropagateIsrChanges(). So if multiple partitions isr change around the same time, that should be only a single child added to the IsrChangeNotificationPath path. Could you explain how IsrChangeNotificationPath accumulates hundreds of children?

lindong28 commented:
@junrao That optimization can reduce the number of IsrChangeNotification znode per broker. But the total number of IsrChangeNotification znode will be proportional to the number of brokers in the cluster. For users who have 100 brokers in a cluster, 100 IsrChangeNotification znodes can be created at a time and controller will need to read znode 10000 times due to this issue.

I observed this when I try to transform a cluster from one-broker-per-disk to one-broker-per-node setup. The cluster has 15 machine where each machine has 10 disks. Thus initially there will be 150 brokers in the one-broker-per-disk setup. It takes really a long time to reassign partitions from all 10 brokers on a given machine to the first broker on this machine due to this issue.

junrao commented:
@lindong28 : Thanks for the explanation. If the cluster is not overloaded, the handling of 1 isrChange event should take just about 10ms. So, processing 100 such events will just take 10 secs. When doing partition reassignment, it's probably unlikely that all 100 brokers will be updating the isrChange path in ZK within the same 10 sec window, right? If so, we probably won't get into the situation that you described?

If the problem is not common, it's probably not worth patching in 1.0.0. If it is common, are there any other ZK path listeners that need to be patched too? We already patched the broker change listener and now want to patch the isrChange listener. I will just wondering whether there are other ZK paths that may need to be patched similarly.

lindong28 commented:
@junrao Let me give a more accurate estimate of the number of zk read/write needed by controller if 100 IsrNotificationZnode is created by 100 brokers.

1) 100 IsrChangeNotification znodes are added to zookeeper in order
2) 100 IsrChangeNotification events will be added to the controller's event queue. The first event has 1 childern, the second event has 2 children.. the 100th event has 100 children.
3) In order to process n'th IsrChangeNotification event, where 1<=n<=100
    - Controller will read zookeeper n times, once for each IsrChangeNotification znode recorded in the event. Note that most of these znodes would have been deleted already when controller processes the previous IsrChangeNotification event.
    - Controller will call `deletePath` n times to delete all n IsrChangeNotification znodes recorded in the event. Similarly, most of these `deletePath` will not be useful because the corresponding znode would have been deleted already.
    - At least one znode will be truely deleted from zookeeper, this will generate another IsrChangeNotification event which has the remaining `100-n` znodes. This will event requires another `200 - 2n` zk read/write operation to be processed by controller.

So the n'th event requires n + n + 200 - 2n = 200 zk read/write. 100 IsrChangeNotification events requires 20000 zk read/write to be processed. If each zk read takes 10 ms, this will take 200 seconds, which is longer than the default IsrChangePropagationInterval 60 seconds. More IsrChangeNotification znodes will be created by brokers before controller finishes processing the previously created IsrChangeNotification, which makes the situation even worse.

Does this explanation make sense?

I will double check if there is any other ZK path listeners that need to be patched. I haven't not found other ZK path yet based on my previous experiments in the test cluster.

lindong28 committed:
Remove log statements and add comment

lindong28 committed:
Merge changes from KAFKA-5877

lindong28 commented:
I have removed the unnecessary log statements from this patch to simplify the review for this patch and make it ready for 1.0 release. I moved the change fro KAFKA-5877 to this patch to hopefully reduce the number of pull requests and commits you need to make for 1.0 release.

junrao commented:
@lindong28 : Thanks for the explanation. LGTM
